staging/src/k8s.io/apiserver/pkg/authentication/authenticator/interfaces.go

```diff
@@ -29,6 +30,71 @@ type Token interface {
 	AuthenticateToken(ctx context.Context, token string) (*Response, bool, error)
 }
 
+// WrapInInstrumentedToken wraps the input Token in the InstrumentedRequest.
+func WrapInInstrumentedToken(token Token, authenticatorName string) *InstrumentedToken {
+	return &InstrumentedToken{AuthenticatorName: authenticatorName, delegate: token}
+}
+
+// InstrumentedToken wraps Token and allows instrumentation to identify authenticators via the AuthenticatorName field.
+type InstrumentedToken struct {
+	// AuthenticatorName identifies the delegate authenticator.
+	AuthenticatorName string
+	delegate          Token
+}
+
+// AuthenticateToken wraps AuthenticateToken of the delegate Token.
+func (i *InstrumentedToken) AuthenticateToken(ctx context.Context, token string) (*Response, bool, error) {
+	authenticationStart := time.Now()
+	resp, ok, err := i.delegate.AuthenticateToken(ctx, token)
+	defer recordAuthMetrics(ctx, resp, ok, err, authenticationStart)
+	if err != nil {
+		err = InstrumentedError{AuthenticatorName: i.AuthenticatorName, error: err}
+	}
+
+	if resp != nil && resp.AuthenticatorName == "" {
+		resp.AuthenticatorName = i.AuthenticatorName
+	}
+
+	return resp, ok, err
+}
+
+// WrapInInstrumentedRequest wraps the input Request in the InstrumentedRequest.
+func WrapInInstrumentedRequest(req Request, authenticatorName string) *InstrumentedRequest {
+	return &InstrumentedRequest{AuthenticatorName: authenticatorName, delegate: req}
+}
+
+// InstrumentedRequest wraps Request and allows instrumentation to identify authenticators via the Name field.
+type InstrumentedRequest struct {
+	// AuthenticatorName identifies the delegate authenticator.
+	AuthenticatorName string
+	delegate          Request
+}
+
+// AuthenticateRequest wraps AuthenticateRequest of the delegate Request.
+func (i *InstrumentedRequest) AuthenticateRequest(req *http.Request) (*Response, bool, error) {
+	authenticationStart := time.Now()
+	resp, ok, err := i.delegate.AuthenticateRequest(req)
+	defer recordAuthMetrics(req.Context(), resp, ok, err, authenticationStart)
+	if err != nil {
+		err = InstrumentedError{AuthenticatorName: i.AuthenticatorName, error: err}
+	}
+
+	// TODO(immutableT) I don't believe we could ever receive a response with AuthenticatorName already set.
```

Comment:
I don't think this should be a todo - Just make it a regular comment, and log something if the assumption is violated.

Reply:
Done.